src/r_builtins/paste.rs

```diff
@@ -0,0 +1,360 @@
+                )))
+            }
+            _ => continue,
+        }
+    }
+
+    // Coerce everything into strings
+    let char_vals: Vec<R> = stack_vals
+        .iter()
+        .map(|(_, v)| v.clone().as_character().unwrap())
+        .collect();
+
+    let vec_of_vectors: Vec<_> = char_vals
+        .iter()
+        .map(|v| v.get_vec_string())
+        // Filtering out Null values
+        .filter(|v| v.len() != 0)
+        .collect();
+
+    // Need the length of longest vector to create an empty vector that others
+    // will go through and re-cycle values as needed
+    let n = vec_of_vectors
+        .iter()
+        .max_by_key(|x| x.len())
+        .unwrap_or(&vec![])
+        .len();
+
+    let mut output = vec!["".to_string(); n];
+
+    for i in 0..vec_of_vectors.len() {
+        output = output
+            .iter()
+            // Any shorter vector will re-cycle its values to the length of
+            // longest one
+            .zip(vec_of_vectors[i].iter().cycle())
+            .map(|(x, y)| {
+                // No need for a sep in the beginning
+                if i == 0 {
+                    return y.clone();
+                }
+                format!("{x}{sep}{y}")
+            })
+            .collect();
+    }
+
+    if collapse_is_specified.is_some() {
+        output = vec![output.join(&collapse)];
+    }
+
+    Ok(R::Vector(output.into()))
+}
+
+#[cfg(test)]
+mod test_primitive_paste {
+    use super::*;
+
+    #[test]
+    fn test_primitive_paste_01() {
+        let mut env = Environment::default();
+
+        // Making a value of args parameter of primitive_paste corresponding to
+        // R c(NULL)
+        let args = ExprList {
+            keys: vec![None],
+            values: vec![Expr::Null],
+        };
+
+        let observed = primitive_paste(args, &mut env).unwrap().get_vec_string();
+        let expected: Vec<String> = vec![];
+
+        assert_eq!(observed, expected);
+    }
+
+    #[test]
+    fn test_primitive_paste_02() {
+        let mut env = Environment::default();
+
+        // Making a value of args parameter of primitive_paste corresponding to
+        // R c(1.1, 2, FALSE, "a", c(1, 2), sep = "+")
+        let mut keys = vec![None; 6];
+        keys.push(Some("sep".to_string()));
+        let args = ExprList {
+            keys,
+            values: vec![
+                Expr::Number(1.1),
+                Expr::Null,
+                Expr::Integer(2),
+                Expr::Bool(false),
+                Expr::String("a".to_string()),
+                Expr::Call(
+                    Box::new(Expr::Symbol("c".to_string())),
+                    ExprList {
+                        keys: vec![None; 2],
+                        values: vec![Expr::Number(1.0), Expr::Number(2.0)],
+                    },
+                ),
+                // sep parameter
+                Expr::String("+".to_string()),
+            ],
+        };
```

Comment:
Allowing parsing will definitely make this a lot easier. If we get further into testing, I think we should add something like

```rust
let args: ExprList = parse("1.1, null, 2, false, \"a\", c(1.0, 2 = 2.0), sep = \"+\"");
```

Reply:
Yeah, I didn't mean to say that I think it will work as-is, but I think we might want to add parsers for these sorts of partial expressions. I think we could have something like (similar to [this parse function](https://github.com/dgkf/R/blob/3b4ecad8d914cea55c0899abf408eabd9954aa6f/src/parser.rs#L39C18-L39C18))

```rust
pub fn parse_args(s: &str) -> Result<ExprList, RError> {
    match RParser::parse(Rule::repl, s) {
        Ok(pairs) => Ok(parse_pairlist(pairs)),
        Err(e) => Err(RError::ParseFailureVerbose(e)),
    }
}
```

Which will return an `ExprList` directly. One of the perks of having access to all the internal parsers at each stage of the syntax tree is that we can expose these helpers for testing if we need them.

Reply:
Thanks so much Doug 🙏 
This worked very nicely and thanks for enabling single quotes 💯